test/integration/storageversion/storage_version_crd_test.go

```diff
@@ -0,0 +1,325 @@
+			return false, nil
+		}
+		if storageVersion.EncodingVersion != expectedNewVersion {
+			lastErr = fmt.Errorf("unexpected encoding version, expected: %v, got: %v", expectedNewVersion, storageVersion.EncodingVersion)
+			return false, nil
+		}
+		return true, nil
+	}); err != nil {
+		t.Fatalf("failed to wait for storage version update: %v, last error: %v", err, lastErr)
+	}
+
+	// cleanup
+	if err := fixtures.DeleteV1CustomResourceDefinition(crd, crdClient); err != nil {
+		t.Fatal(err)
+	}
+	if err := kubeclient.InternalV1alpha1().StorageVersions().Delete(context.TODO(), gr, metav1.DeleteOptions{}); err != nil {
+		t.Errorf("failed to delete storage version: %v", err)
+	}
+}
+
+func TestStorageVersionMultipleCRDs(t *testing.T) {
+	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.StorageVersionAPI, true)()
+	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.APIServerIdentity, true)()
+	etcdConfig := framework.SharedEtcd()
+	server := kubeapiservertesting.StartTestServerOrDie(t, nil,
+		[]string{
+			"--runtime-config=internal.apiserver.k8s.io/v1alpha1=true",
+		},
+		etcdConfig)
+	crdClient := apiextensionsclientset.NewForConfigOrDie(server.ClientConfig)
+	crd := etcd.GetCustomResourceDefinitionData()[0]
+	etcd.CreateTestCRDs(t, crdClient, false, crd)
+	defer server.TearDownFn()
+	dynamicClient, err := dynamic.NewForConfig(server.ClientConfig)
+	if err != nil {
+		t.Fatalf("unexpected error creating dynamic client: %v", err)
+	}
+	gvr := schema.GroupVersionResource{Group: crd.Spec.Group, Version: crd.Spec.Versions[0].Name, Resource: crd.Spec.Names.Plural}
+	errCh := make(chan error)
+	// keep flipping the storage version of the CRD and create new CRs
+	go func() {
+		for i := 0; i < 10; i++ {
+			createdCRD, err := crdClient.ApiextensionsV1().CustomResourceDefinitions().Get(context.TODO(), crd.Name, metav1.GetOptions{})
+			if err != nil {
+				errCh <- fmt.Errorf("failed to get crd: %w", err)
+				return
+			}
+
+			// add a new version to the CRD and set it as the storage version
+			newVersion := createdCRD.Spec.Versions[i]
+			newVersion.Name = fmt.Sprintf("v%d", i+2)
+			createdCRD.Spec.Versions[i].Storage = false
+			createdCRD.Spec.Versions = append(createdCRD.Spec.Versions, newVersion)
+			if _, err := crdClient.ApiextensionsV1().CustomResourceDefinitions().Update(context.TODO(), createdCRD, metav1.UpdateOptions{}); err != nil {
+				errCh <- fmt.Errorf("failed to update crd: %w", err)
+				return
+			}
+			cr := &unstructured.Unstructured{}
+			cr.SetName(newVersion.Name)
+			cr.SetAPIVersion(fmt.Sprintf("%s/v1", crd.Spec.Group))
+			cr.SetKind(crd.Spec.Names.Kind)
+			if _, err = dynamicClient.Resource(gvr).Namespace("default").Create(context.TODO(), cr, metav1.CreateOptions{}); err != nil {
+				errCh <- fmt.Errorf("unexpected error creating cr for version %v: %w", newVersion.Name, err)
+				return
+			}
```

Comment:
Create CRs in a separate go routine?  By the time update returns there is ample time for storage handlers to be fully swapped out. Interleaving operations is more probable if the CR writes are in parallel.

Reply:
Done